<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

¡Hola Gabriela!

Mi nombre es <b>Juan Tellez</b>. Un gusto conocerte, seré tu revisor en este proyecto.

Cuando encuentre un error por primera vez, simplemente lo señalaré para que puedas identificarlo y corregirlo por tu cuenta. Además, a lo largo del notebook iré dejando algunas observaciones para mejorar tu código, fortalecer tu análisis y comentar también tus interpretaciones sobre el tema.

Si en esta primera orientación aún no logras resolver algún punto, en la próxima iteración te dejaré una pista más precisa y, cuando sea necesario, algunos ejemplos prácticos. La idea es que este proceso no sea solo una corrección, sino también una oportunidad para que desarrolles criterio como analista de datos o científica de datos.

También estaré abierto a comentarios y discusiones sobre el tema.

Encontrarás mis comentarios a continuación. Por favor, no los muevas, modifiques ni elimines.

Puedes encontrar mis comentarios en cuadros verdes, amarillos o rojos como estos:

</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Éxito. Todo se ha hecho de forma exitosa.

</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Observación. Algunas recomendaciones o ajustes no bloqueantes.

</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Necesita arreglos. Este apartado necesita algunas correcciones. El trabajo no puede ser aceptado con comentarios rojos.

</div>

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a><br><br>


</div>

<div class="alert alert-block alert-success">
<b>Review General. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Gabriela, quería tomarme este espacio al inicio de tu proyecto para comentarte mis apreciaciones generales de esta primera iteración.

Siempre me gusta comenzar dando la bienvenida al mundo de los datos a los estudiantes. Te deseo lo mejor en este camino y espero que consigas lograr tus objetivos. Personalmente me gusta compartir este consejo: está bien equivocarse. Es normal y, de hecho, es una de las mejores formas de aprender. Los errores te ayudan a descubrir detalles que no siempre se ven a primera vista, y esa experiencia es la que poco a poco te va formando como una mejor analista o científica de datos.

Ahora sí, yendo a esta entrega:

Se nota un esfuerzo importante por desarrollar el proyecto completo y mantener una estructura ordenada. El notebook incluye las etapas principales solicitadas: preparación de datos, rentabilidad, funnel, retención, experimento estadístico y comunicación mediante Power BI. También dejaste resultados ejecutados, nombres descriptivos y comentarios que permiten seguir con facilidad la intención de cada bloque.

Entre las fortalezas principales destaco la limpieza realizada sobre los tres datasets, la conversión de fechas, la revisión de valores inválidos, la corrección de categorías, la exportación de archivos limpios y el cálculo correcto del profit dentro del notebook. También está bien desarrollado el test A/B: planteas las hipótesis, preparas conversiones y usuarios, aplicas el z-test de proporciones e interpretas correctamente el p-valor.

El dashboard presenta un buen esfuerzo visual para esta etapa. Incluye una vista ejecutiva, tarjetas de indicadores, evolución mensual, filtros y una segunda página de detalle. Tu propuesta ya demuestra una intención clara de organizar la información desde el resumen hacia el detalle.

Éxitos Gabriela.

Saludos.
</div>


# Proyecto RappiPlus: de datos a decisiones de negocio

**Introducción**


El objetivo de este proyecto es evaluar el desempeño del servicio **RappiPlus** para apoyar **decisiones de negocio basadas en datos**.

Se trabajan con múltiples datasets del negocio:

- **rappiplus_orders_raw.csv** → información de pedidos, precios, descuentos y revenue  
- **rappiplus_catalog.csv** → costos de productos, categorías y proveedores  
- **rappiplus_marketing_spend.csv** → inversión en marketing por canal y país  
- **events / users / user_activity (SQL)** → comportamiento del usuario dentro de la plataforma  
- **experiment_checkout_ui.csv** → resultados de un experimento A/B en el checkout  

El análisis sigue una lógica clara y progresiva:

1. 🔍 Evaluar si podemos confiar en los datos (calidad de datos en Python) 

2. 💰 Analizar si el negocio es rentable (revenue, costos y profit)  

3. 🛒 Entender dónde se pierden los usuarios (funnel de conversión)  

4. 🔁 Evaluar si los usuarios regresan (retención por cohortes)  

5. 🧪 Validar si los cambios generan impacto (test estadístico)  

6. 📊 Comunicar los resultados (dashboard en BI)  

A lo largo del proyecto, se transforman datos en insights para responder preguntas clave del negocio y proponer **recomendaciones accionables**.

---

## 🔹 Paso 1: Cargar y validar la calidad de los datos

---

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:** Familiarizarte con la estructura de los datasets del negocio antes de analizarlos.

**Instrucciones:**

- Importa las librerías necesarias
- Carga los archivos:
  - `rappiplus_orders_raw.csv`
  - `rappiplus_catalog.csv`
  - `rappiplus_marketing_spend.csv`
- Guarda los DataFrames en:
  - `orders`, `catalog`, `marketing`
- Explora cada dataset.

---

In [10]:
# importar librerías
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [11]:
# cargar archivos
orders = pd.read_csv('rappiplus_orders_raw.csv')
catalog = pd.read_csv('rappiplus_catalog.csv')
marketing =pd.read_csv('rappiplus_marketing_spend.csv')

In [12]:
# explorar datasets ORDERS
orders.head(5)

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28


In [13]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_pedido           25100 non-null  object 
 1   id_usuario          25100 non-null  object 
 2   fecha_hora_pedido   25100 non-null  object 
 3   pais                24800 non-null  object 
 4   dispositivo         25080 non-null  object 
 5   fuente_referencia   25070 non-null  object 
 6   nombre_producto     25070 non-null  object 
 7   categoria_producto  25020 non-null  object 
 8   cantidad            25050 non-null  float64
 9   precio_unitario     25050 non-null  float64
 10  monto_descuento     25050 non-null  float64
 11  monto_total         25100 non-null  float64
dtypes: float64(4), object(8)
memory usage: 2.3+ MB


In [14]:
# explorar datasets CATALOG
catalog.head(12)

,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electrónica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electrónica,10.12,King Ltd
2,Tablet-Standard-64GB,Electrónica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"
5,Sneakers-Urban-42,Moda,17.21,Greene-Smith
6,Jacket-Winter-M,Moda,189.31,Mcmillan-Rhodes


In [15]:
catalog.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   nombre_producto     7 non-null      object 
 1   categoria_producto  7 non-null      object 
 2   costo_unitario      7 non-null      float64
 3   proveedor           7 non-null      object 
dtypes: float64(1), object(3)
memory usage: 352.0+ bytes


In [16]:
# explorar datasets MARKETING
marketing.head(5)

,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40


In [17]:
marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   fecha       1620 non-null   object 
 1   pais        1620 non-null   object 
 2   id_campaña  1620 non-null   object 
 3   canal       1519 non-null   object 
 4   gasto       1620 non-null   float64
dtypes: float64(1), object(4)
memory usage: 63.4+ KB


---

### Revisión y calidad de datos

**🎯 Objetivo:** Detectar y corregir problemas en los datos que puedan afectar el análisis de revenue, costos y rentabilidad.

Se revisan los 3 datasets
- Validar y convertir fechas al formato correcto  
- Revisar variables numéricas (sin negativos o ceros inválidos)  
- Verificar consistencia de montos  
- Eliminar duplicados  
- Revisar variables categóricas 

---

In [18]:
#CREACION DE COPIA DE TRABAJO
orders_work = orders.copy()
catalog_work=catalog.copy()
marketing_work=marketing.copy()

In [19]:
# REVISION Y CALIDAD DE DATOS DE ORDERS

#Correccion de formato de fecha
orders_work['fecha_hora_pedido'] = pd.to_datetime(orders_work['fecha_hora_pedido'])
orders_work.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   id_pedido           25100 non-null  object        
 1   id_usuario          25100 non-null  object        
 2   fecha_hora_pedido   25100 non-null  datetime64[ns]
 3   pais                24800 non-null  object        
 4   dispositivo         25080 non-null  object        
 5   fuente_referencia   25070 non-null  object        
 6   nombre_producto     25070 non-null  object        
 7   categoria_producto  25020 non-null  object        
 8   cantidad            25050 non-null  float64       
 9   precio_unitario     25050 non-null  float64       
 10  monto_descuento     25050 non-null  float64       
 11  monto_total         25100 non-null  float64       
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 2.3+ MB


In [20]:
#Conteo de NA por columna
orders_work.isna().sum()

id_pedido               0
id_usuario              0
fecha_hora_pedido       0
pais                  300
dispositivo            20
fuente_referencia      30
nombre_producto        30
categoria_producto     80
cantidad               50
precio_unitario        50
monto_descuento        50
monto_total             0
dtype: int64

In [21]:
#Conteo de NA en total del data set Orders
orders_work.isna().sum().sum()

610

In [22]:
#Revision de columnas que tienen valores ausentes
orders_work.columns[orders_work.isna().any()]

Index(['pais', 'dispositivo', 'fuente_referencia', 'nombre_producto',
       'categoria_producto', 'cantidad', 'precio_unitario', 'monto_descuento'],
      dtype='object')

In [23]:
# Porcentaje de valores ausentes por columna
porcentaje_na = (orders_work.isna().sum() / len(orders_work)) * 100
porcentaje_na = porcentaje_na.round(2)

# Ordenado de mayor a menor para priorizar
print(porcentaje_na.sort_values(ascending=False))

pais                  1.20
categoria_producto    0.32
cantidad              0.20
precio_unitario       0.20
monto_descuento       0.20
fuente_referencia     0.12
nombre_producto       0.12
dispositivo           0.08
id_pedido             0.00
id_usuario            0.00
fecha_hora_pedido     0.00
monto_total           0.00
dtype: float64


In [24]:
#Se observa que el % de valores ausentes el mayor de 1.20 % y todos los demas son menores, por lo que se toma la decision de va a reemplazar con la moda y la media

In [25]:
# Categóricas → moda
orders_work['pais'] = orders_work['pais'].fillna(orders_work['pais'].mode()[0])
orders_work['dispositivo'] = orders_work['dispositivo'].fillna(orders_work['dispositivo'].mode()[0])
orders_work['fuente_referencia'] = orders_work['fuente_referencia'].fillna(orders_work['fuente_referencia'].mode()[0])
orders_work['nombre_producto'] = orders_work['nombre_producto'].fillna(orders_work['nombre_producto'].mode()[0])
orders_work['categoria_producto'] = orders_work['categoria_producto'].fillna(orders_work['categoria_producto'].mode()[0])

# Numéricas → media
orders_work['cantidad'] = orders_work['cantidad'].fillna(orders_work['cantidad'].mean())
orders_work['precio_unitario'] = orders_work['precio_unitario'].fillna(orders_work['precio_unitario'].mean())
orders_work['monto_descuento'] = orders_work['monto_descuento'].fillna(orders_work['monto_descuento'].mean())

print(orders_work.isna().sum())

id_pedido             0
id_usuario            0
fecha_hora_pedido     0
pais                  0
dispositivo           0
fuente_referencia     0
nombre_producto       0
categoria_producto    0
cantidad              0
precio_unitario       0
monto_descuento       0
monto_total           0
dtype: int64


<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Realizaste un tratamiento explícito de los valores ausentes, lo cual es positivo. Sin embargo, conviene revisar con un poco más de detalle la imputación de <code>cantidad</code>, <code>nombre_producto</code> y <code>categoria_producto</code>.

En esta sección, la media inicial de <code>cantidad</code> está afectada por valores extremos y posteriormente las filas imputadas pueden terminar eliminándose con el filtro del IQR. Además, completar producto y categoría de manera independiente con sus respectivas modas podría crear combinaciones que no existían originalmente.

En un escenario profesional, estas decisiones deben mantener la coherencia entre producto, categoría, cantidad y monto. Para fortalecer la limpieza, revisa cuántas filas son afectadas por cada imputación y decide si conviene recuperar el dato mediante otra columna relacionada, conservar una categoría como “desconocido” o retirar únicamente los registros que no puedan validarse.

</div>

In [26]:
orders_work.describe()

,cantidad,precio_unitario,monto_descuento,monto_total
count,25100.000000,25100.000000,25100.000000,2.510000e+04
mean,7.092735,259.305549,4.500798,2.072680e+03
std,295.981747,138.588213,5.217805,9.894995e+04
min,-2.000000,20.030000,0.000000,-4.926500e+02
25%,1.000000,138.592500,0.000000,1.805075e+02
50%,2.000000,259.305549,0.000000,3.417500e+02
75%,2.000000,380.092500,10.000000,5.185800e+02
max,20000.000000,499.960000,15.000000,8.840200e+06


In [27]:
#Revision continua de limpieza de datos por valores que son inconsistentes 

neg_cant = (orders_work['cantidad'] < 0).sum()
cero_cant = (orders_work['cantidad'] == 0).sum()
print(f"cantidad negativa: {neg_cant} | cantidad cero: {cero_cant}")

Q1 = orders_work['cantidad'].quantile(0.25)
Q3 = orders_work['cantidad'].quantile(0.75)
IQR = Q3 - Q1
limite_superior = Q3 + 3 * IQR
print(f"Límite superior IQR para cantidad: {limite_superior}")

cantidad negativa: 4 | cantidad cero: 0
Límite superior IQR para cantidad: 5.0


In [28]:
# Aplicar filtro de cantidad
filas_antes = len(orders_work)

orders_work = orders_work[
    (orders_work['cantidad'] > 0) &
    (orders_work['cantidad'] <= limite_superior)
]

filas_despues = len(orders_work)
print(f"Filas eliminadas por cantidad: {filas_antes - filas_despues}")
print(orders_work['cantidad'].describe())

Filas eliminadas por cantidad: 64
count    25036.000000
mean         1.504953
std          0.499985
min          1.000000
25%          1.000000
50%          2.000000
75%          2.000000
max          2.000000
Name: cantidad, dtype: float64


In [29]:
print("Filas con cantidad > 5:", (orders_work['cantidad'] > 5).sum())
print(orders_work[orders_work['cantidad'] > 5]['cantidad'].value_counts().sort_index())

Filas con cantidad > 5: 0
Series([], Name: cantidad, dtype: int64)


In [30]:
# Diagnóstico de negativos
neg_monto = (orders_work['monto_total'] < 0).sum()
print(f"monto_total negativo: {neg_monto}")
print(orders_work[orders_work['monto_total'] < 0][['id_pedido','cantidad','precio_unitario','monto_descuento','monto_total']].head(10))

# Diagnóstico de outliers extremos
p99 = orders_work['monto_total'].quantile(0.99)
p999 = orders_work['monto_total'].quantile(0.999)
print(f"Percentil 99: {p99}")
print(f"Percentil 99.9: {p999}")
print(f"Máximo actual: {orders_work['monto_total'].max()}")

# Revisa cuántas filas quedan por encima de distintos cortes
print("Filas > p99*3:", (orders_work['monto_total'] > p99*3).sum())
print("Filas > 10000:", (orders_work['monto_total'] > 10000).sum())

monto_total negativo: 0
Empty DataFrame
Columns: [id_pedido, cantidad, precio_unitario, monto_descuento, monto_total]
Index: []
Percentil 99: 977.393
Percentil 99.9: 996.6282500000002
Máximo actual: 999.89
Filas > p99*3: 0
Filas > 10000: 0


In [31]:
# Consistencia aritmética
orders_work['monto_calculado'] = (
    orders_work['cantidad'] * orders_work['precio_unitario']
    - orders_work['monto_descuento']
)
orders_work['diff_monto'] = (
    orders_work['monto_total'] - orders_work['monto_calculado']
).abs()

inconsistentes = orders_work[orders_work['diff_monto'] > 1]  # tolerancia de redondeo
print(f"Filas con inconsistencia aritmética: {len(inconsistentes)}")
print(inconsistentes[['id_pedido','cantidad','precio_unitario','monto_descuento','monto_total','monto_calculado','diff_monto']].head(10))
print(orders_work['diff_monto'].describe())

Filas con inconsistencia aritmética: 0
Empty DataFrame
Columns: [id_pedido, cantidad, precio_unitario, monto_descuento, monto_total, monto_calculado, diff_monto]
Index: []
count    2.503600e+04
mean     2.498802e-03
std      4.329521e-03
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      5.684342e-14
max      1.000000e-02
Name: diff_monto, dtype: float64


In [32]:
# Cierre del pipeline de limpieza
orders_work = orders_work.drop(columns=['monto_calculado', 'diff_monto'])

print(f"Dataset final: {orders_work.shape[0]} filas, {orders_work.shape[1]} columnas")
print(orders_work.describe())

Dataset final: 25036 filas, 12 columnas
           cantidad  precio_unitario  monto_descuento   monto_total
count  25036.000000     25036.000000     25036.000000  25036.000000
mean       1.504953       259.293912         4.502516    385.807512
std        0.499985       138.716939         5.223500    255.750599
min        1.000000        20.030000         0.000000      5.240000
25%        1.000000       138.392500         0.000000    180.330000
50%        2.000000       258.700000         0.000000    341.360000
75%        2.000000       380.297500        10.000000    517.640000
max        2.000000       499.960000        15.000000    999.890000


In [33]:
# PASO VERIFICACION 1: Eliminar columnas auxiliares de validación
orders_work = orders_work.drop(columns=['monto_calculado', 'diff_monto'], errors='ignore')

print(orders_work.columns.tolist())

['id_pedido', 'id_usuario', 'fecha_hora_pedido', 'pais', 'dispositivo', 'fuente_referencia', 'nombre_producto', 'categoria_producto', 'cantidad', 'precio_unitario', 'monto_descuento', 'monto_total']


In [34]:
# =========================================================
# PASO 2: Diagnóstico de 'pais' antes de corregir
# =========================================================
print("Valores únicos ANTES:", orders_work['pais'].unique())
print("Conteo por valor:")
print(orders_work['pais'].value_counts())

Valores únicos ANTES: ['Argentina' 'Mexico' 'Colombia' 'mexico' 'colombia' 'argentina']
Conteo por valor:
Colombia     7800
Mexico       7485
Argentina    7268
mexico        863
colombia      822
argentina     798
Name: pais, dtype: int64


In [35]:
# =========================================================
# PASO 3: Normalizar 'pais'
# =========================================================
orders_work['pais'] = orders_work['pais'].str.strip().str.title()

print("Valores únicos DESPUÉS:", orders_work['pais'].unique())
assert orders_work['pais'].nunique() == 3, "Aún hay variantes sin normalizar"

Valores únicos DESPUÉS: ['Argentina' 'Mexico' 'Colombia']


In [36]:
# =========================================================
# PASO 4: Revisar TODAS las columnas categóricas (no solo 'pais')
# =========================================================
cols_categoricas = ['pais', 'dispositivo', 'fuente_referencia', 'categoria_producto']

for col in cols_categoricas:
    print(f"--- {col} ---")
    print(orders_work[col].unique())

--- pais ---
['Argentina' 'Mexico' 'Colombia']
--- dispositivo ---
['desktop' 'mobile']
--- fuente_referencia ---
['organic' 'paid_search' 'social']
--- categoria_producto ---
['Moda' 'Electronica' 'Hogar']


In [37]:
# =========================================================
# PASO 6: Verificación final antes de exportar
# =========================================================
print(f"Dataset final: {orders_work.shape[0]} filas, {orders_work.shape[1]} columnas")
print(orders_work.columns.tolist())
for col in cols_categoricas:
    print(col, "->", orders_work[col].nunique(), "valores únicos")

Dataset final: 25036 filas, 12 columnas
['id_pedido', 'id_usuario', 'fecha_hora_pedido', 'pais', 'dispositivo', 'fuente_referencia', 'nombre_producto', 'categoria_producto', 'cantidad', 'precio_unitario', 'monto_descuento', 'monto_total']
pais -> 3 valores únicos
dispositivo -> 2 valores únicos
fuente_referencia -> 3 valores únicos
categoria_producto -> 3 valores únicos


In [38]:
# =========================================================
# FIX: Corregir tilde en categoria_producto
# =========================================================
print("Antes:", orders_work['categoria_producto'].unique())

orders_work['categoria_producto'] = orders_work['categoria_producto'].replace({
    'Electronica': 'Electrónica'
})

print("Después:", orders_work['categoria_producto'].unique())

Antes: ['Moda' 'Electronica' 'Hogar']
Después: ['Moda' 'Electrónica' 'Hogar']


In [39]:
# REVISION Y CALIDAD DE DATOS DE CATALOG
catalog_work.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   nombre_producto     7 non-null      object 
 1   categoria_producto  7 non-null      object 
 2   costo_unitario      7 non-null      float64
 3   proveedor           7 non-null      object 
dtypes: float64(1), object(3)
memory usage: 352.0+ bytes


In [40]:
# Duplicados de nombre_producto (no debería repetirse si es un catálogo maestro)
print("Duplicados en nombre_producto:", catalog_work['nombre_producto'].duplicated().sum())

# Costo_unitario inválido (negativo o cero)
print(catalog_work[catalog_work['costo_unitario'] <= 0])

# Revisar inconsistencias de texto (espacios extra, mayúsculas/minúsculas distintas para el mismo valor)
print(catalog_work['nombre_producto'].unique())
print(catalog_work['categoria_producto'].unique())
print(catalog_work['proveedor'].unique())

Duplicados en nombre_producto: 0
Empty DataFrame
Columns: [nombre_producto, categoria_producto, costo_unitario, proveedor]
Index: []
['Laptop-Gaming-16GB' 'Phone-Pro-128GB' 'Tablet-Standard-64GB'
 'Blender-XL-Red' 'Vacuum-Pro-Black' 'Sneakers-Urban-42' 'Jacket-Winter-M']
['Electrónica' 'Hogar' 'Moda']
['Fuller, Pena and Myers' 'King Ltd' 'Bowers LLC' 'Long-Reid'
 'Rivera, Carr and Finley' 'Greene-Smith' 'Mcmillan-Rhodes']


In [41]:
catalog_work.describe()

,costo_unitario
count,7.000000
mean,102.252857
std,111.011563
min,10.120000
25%,16.905000
50%,25.210000
75%,182.975000
max,280.680000


In [42]:
# =========================================================
# PASO 1: Nulos y duplicados
# =========================================================
print(catalog_work.isnull().sum())
print("Filas duplicadas (exactas):", catalog_work.duplicated().sum())
print("nombre_producto duplicado:", catalog_work.duplicated(subset=['nombre_producto']).sum())

catalog_work = catalog_work.drop_duplicates(subset=['nombre_producto'], keep='first')

# =========================================================
# PASO 2: Normalización de texto (espacios, mayúsculas, tildes ocultas)
# =========================================================
cols_texto = ['nombre_producto', 'categoria_producto', 'proveedor']

for col in cols_texto:
    catalog_work[col] = catalog_work[col].str.strip()  # elimina espacios al inicio/final

# Revisa si hay variantes de la misma categoría por typo/case
print(catalog_work['categoria_producto'].unique())
print(catalog_work['proveedor'].unique())

# =========================================================
# PASO 3: Validación de tipo de dato numérico
# =========================================================
print(catalog_work['costo_unitario'].dtype)
print(catalog_work['costo_unitario'].isnull().sum())

# =========================================================
# PASO 4: Valores inválidos en costo_unitario (negativos o cero)
# =========================================================
invalidos = catalog_work[catalog_work['costo_unitario'] <= 0]
print(f"Filas con costo_unitario <= 0: {len(invalidos)}")
print(invalidos)

catalog_work = catalog_work[catalog_work['costo_unitario'] > 0]

print(catalog_work.describe())

nombre_producto       0
categoria_producto    0
costo_unitario        0
proveedor             0
dtype: int64
Filas duplicadas (exactas): 0
nombre_producto duplicado: 0
['Electrónica' 'Hogar' 'Moda']
['Fuller, Pena and Myers' 'King Ltd' 'Bowers LLC' 'Long-Reid'
 'Rivera, Carr and Finley' 'Greene-Smith' 'Mcmillan-Rhodes']
float64
0
Filas con costo_unitario <= 0: 0
Empty DataFrame
Columns: [nombre_producto, categoria_producto, costo_unitario, proveedor]
Index: []
       costo_unitario
count        7.000000
mean       102.252857
std        111.011563
min         10.120000
25%         16.905000
50%         25.210000
75%        182.975000
max        280.680000


In [43]:
# REVISION Y CALIDAD DE DATOS DE MARKETING
marketing_work.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   fecha       1620 non-null   object 
 1   pais        1620 non-null   object 
 2   id_campaña  1620 non-null   object 
 3   canal       1519 non-null   object 
 4   gasto       1620 non-null   float64
dtypes: float64(1), object(4)
memory usage: 63.4+ KB


In [44]:
#Correccion de formato de fecha
marketing_work['fecha'] = pd.to_datetime(marketing_work['fecha'])
marketing_work.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   fecha       1620 non-null   datetime64[ns]
 1   pais        1620 non-null   object        
 2   id_campaña  1620 non-null   object        
 3   canal       1519 non-null   object        
 4   gasto       1620 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(3)
memory usage: 63.4+ KB


In [45]:
#Conteo de NA por columna
marketing_work.isna().sum()

fecha           0
pais            0
id_campaña      0
canal         101
gasto           0
dtype: int64

In [46]:
# Porcentaje de Valores Ausentes
porcentaje_na_canal = (marketing_work['canal'].isna().sum() / len(marketing_work)) * 100
print(f"% de valores ausentes en 'canal': {porcentaje_na_canal:.2f}%")

% de valores ausentes en 'canal': 6.23%


In [47]:
#Debido a que es una columna critica, no se va a eliminar, se va a considera analisis de la distribucion

In [48]:
tabla_resumen = pd.DataFrame({
    'frecuencia': marketing_work['canal'].value_counts(dropna=False),
    'porcentaje': (marketing_work['canal'].value_counts(dropna=False, normalize=True) * 100).round(2)
})
print(tabla_resumen)

             frecuencia  porcentaje
paid_search         507       31.30
organic             506       31.23
social              506       31.23
NaN                 101        6.23


In [49]:
marketing_work['canal'] = marketing_work['canal'].fillna(
    marketing_work.apply(lambda row: row['id_campaña'].replace(f"_{row['pais']}", ''), axis=1)
)
# Verificación de consistencia general
# Gasto inválido (negativo o cero)
gasto_invalido = marketing_work[marketing_work['gasto'] <= 0]
print("Filas con gasto <= 0:", len(gasto_invalido))

# Consistencia de texto en categóricas
print("\nValores únicos 'pais':", marketing_work['pais'].unique())
print("Valores únicos 'canal':", marketing_work['canal'].unique())

# Rango de fechas (detectar fechas fuera de lo esperado)
print("\nRango de fechas:", marketing_work['fecha'].min(), "a", marketing_work['fecha'].max())

# Confirmación final
print(marketing_work.isna().sum())
print(f"\nShape final: {marketing_work.shape}")

Filas con gasto <= 0: 0

Valores únicos 'pais': ['Mexico' 'Colombia' 'Argentina']
Valores únicos 'canal': ['organic' 'paid_search' 'social']

Rango de fechas: 2025-01-01 00:00:00 a 2025-06-29 00:00:00
fecha         0
pais          0
id_campaña    0
canal         0
gasto         0
dtype: int64

Shape final: (1620, 5)


In [50]:
marketing_work.describe()

,gasto
count,1620.00000
mean,1772.74292
std,734.43294
min,501.11000
25%,1128.03000
50%,1782.42500
75%,2420.68500
max,2999.36000


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Buen trabajo con la preparación de los tres datasets.

En órdenes convertiste la fecha, revisaste nulos, detectaste cantidades inválidas y valores extremos, validaste la consistencia entre cantidad, precio, descuento y monto total, y normalizaste categorías como <code>Electronica</code> a <code>Electrónica</code>. En catálogo comprobaste la unicidad de los productos y la validez del costo unitario. En marketing reconstruiste correctamente el canal faltante a partir de <code>id_campaña</code> y verificaste que el gasto fuera positivo.

Este orden de trabajo es valioso porque permite construir las métricas financieras sobre datos más confiables y deja una base consistente para el dashboard.

</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

La revisión de duplicados quedó bien documentada para el catálogo, pero no se observa la misma validación explícita para <code>orders_work</code> y <code>marketing_work</code>.

Esto no significa necesariamente que existan duplicados, pero es importante dejar evidencia del diagnóstico. Un pedido repetido puede inflar revenue y cantidades, mientras que una campaña repetida puede aumentar artificialmente el gasto de marketing.

Antes de exportar los archivos limpios, agrega el conteo de duplicados exactos en órdenes y marketing. Para órdenes también puedes comprobar si <code>id_pedido</code> debería ser único. Si no encuentras duplicados, basta con mostrar el resultado y conservar los registros.

</div>

**📦 Exportación**: Una vez finalizada la limpieza, se exportan los datasets para utilizarlos en la última etapa del proyecto.

In [51]:
# exportar datasets
orders_work.to_csv('orders_clean.csv', index=False)
catalog_work.to_csv('catalog_clean.csv', index=False)
marketing_work.to_csv('marketing_clean.csv', index=False)

In [52]:
orders_work.describe()

,cantidad,precio_unitario,monto_descuento,monto_total
count,25036.000000,25036.000000,25036.000000,25036.000000
mean,1.504953,259.293912,4.502516,385.807512
std,0.499985,138.716939,5.223500,255.750599
min,1.000000,20.030000,0.000000,5.240000
25%,1.000000,138.392500,0.000000,180.330000
50%,2.000000,258.700000,0.000000,341.360000
75%,2.000000,380.297500,10.000000,517.640000
max,2.000000,499.960000,15.000000,999.890000


In [53]:
catalog_work.describe()

,costo_unitario
count,7.000000
mean,102.252857
std,111.011563
min,10.120000
25%,16.905000
50%,25.210000
75%,182.975000
max,280.680000


In [54]:
marketing_work.describe()

,gasto
count,1620.00000
mean,1772.74292
std,734.43294
min,501.11000
25%,1128.03000
50%,1782.42500
75%,2420.68500
max,2999.36000


---

## 🔹 Paso 2: Analizar si el negocio es rentable

### 2.1 Cálculo de KPIs principales

**🎯 Objetivo:** Calcular los indicadores clave del negocio para evaluar ingresos, costos y rentabilidad.

Se usan los 3 datasets (`orders`, `catalog`, `marketing_spend`):

**📊 Parte 1: Rentabilidad del negocio**
- ¿Cuál es el ingreso total (revenue)? 
- ¿Cuál es el costo total? 
- ¿Cuánto se ha invertido en marketing? 
- ¿El negocio es rentable? (calcular profit)  

---

**📈 Parte 2: Comportamiento de ventas**
- ¿Cuál es el ticket promedio por orden? 
- ¿Cuál es la cantidad promedio de productos por orden? 
- ¿Cuál es el producto más vendido?
- ¿Cuánto se ha gastado en marketing por canal? 

In [55]:
orders1= pd.read_csv('orders_clean.csv')
catalog1= pd.read_csv('catalog_clean.csv')
marketing1= pd.read_csv('marketing_clean.csv')

In [56]:
# Parte 1: Rentabilidad del negocio

In [58]:
#1. Ingreso total (Revenue)
revenue_total = orders1['monto_total'].sum()
print(f"Ingreso total: ${revenue_total:,.2f}")

Ingreso total: $9,659,076.86


In [59]:
# Merge para traer costo_unitario a cada pedido
orders_con_costo = orders1.merge(
    catalog1[['nombre_producto', 'costo_unitario']],
    on='nombre_producto',
    how='left'
)
# Verificar que el merge no generó NaN (productos sin match en el catálogo)
print("Pedidos sin costo_unitario tras el merge:", orders_con_costo['costo_unitario'].isna().sum())

# Costo total = costo_unitario * cantidad, sumado
orders_con_costo['costo_total_linea'] = orders_con_costo['costo_unitario'] * orders_con_costo['cantidad']
costo_total = orders_con_costo['costo_total_linea'].sum()
print(f"Costo total: ${costo_total:,.2f}")

Pedidos sin costo_unitario tras el merge: 0
Costo total: $3,846,218.06


In [60]:
# Inversión total en marketing
marketing_total = marketing1['gasto'].sum()
print(f"Inversión total en marketing: ${marketing_total:,.2f}")

Inversión total en marketing: $2,871,843.53


In [61]:
# Profit y rentabilidad
profit = revenue_total - costo_total - marketing_total
margen_pct = (profit / revenue_total) * 100

print(f"Profit: ${profit:,.2f}")
print(f"Margen de rentabilidad: {margen_pct:.2f}%")
print("¿Es rentable?:", "Sí" if profit > 0 else "No")

Profit: $2,941,015.27
Margen de rentabilidad: 30.45%
¿Es rentable?: Sí


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Muy buen desarrollo de la sección de rentabilidad.

El merge entre órdenes y catálogo se realiza mediante <code>nombre_producto</code>, verificas que no queden pedidos sin costo y calculas el costo total como <code>costo_unitario * cantidad</code>. Posteriormente defines el profit como revenue menos costos y menos marketing, obteniendo aproximadamente <b>$2.94 millones</b> y un margen de <b>30.45%</b>.

Esta fórmula está alineada con la rúbrica y permite concluir correctamente que el negocio es rentable después de considerar tanto el costo de los productos como la inversión de marketing.

</div>

In [62]:
#2. Costo total
# Ticket promedio por orden
ticket_promedio = orders1['monto_total'].mean()
print(f"Ticket promedio por orden: ${ticket_promedio:,.2f}")

Ticket promedio por orden: $385.81


In [63]:
# Cantidad promedio de productos por orden
cantidad_promedio = orders1['cantidad'].mean()
print(f"Cantidad promedio de productos por orden: {cantidad_promedio:.2f}")

Cantidad promedio de productos por orden: 1.50


In [64]:
# Producto mas vendido

# Por unidades vendidas (suma de cantidad)
producto_mas_vendido_unidades = orders1.groupby('nombre_producto')['cantidad'].sum().sort_values(ascending=False)
print("Producto más vendido por unidades:")
print(producto_mas_vendido_unidades.head(5))

# Por número de órdenes en que aparece (frecuencia)
producto_mas_vendido_ordenes = orders1['nombre_producto'].value_counts()
print("\nProducto más vendido por número de órdenes:")
print(producto_mas_vendido_ordenes.head(5))

# Por ingreso generado (monto_total)
producto_mas_vendido_ingreso = orders1.groupby('nombre_producto')['monto_total'].sum().sort_values(ascending=False)
print("\nProducto con mayor ingreso generado:")
print(producto_mas_vendido_ingreso.head(5))

Producto más vendido por unidades:
nombre_producto
Vacuum-Pro-Black      6362.0
Blender-XL-Red        6305.0
Jacket-Winter-M       6281.0
Sneakers-Urban-42     6190.0
Laptop-Gaming-16GB    4219.0
Name: cantidad, dtype: float64

Producto más vendido por número de órdenes:
Vacuum-Pro-Black      4223
Blender-XL-Red        4192
Jacket-Winter-M       4182
Sneakers-Urban-42     4141
Laptop-Gaming-16GB    2780
Name: nombre_producto, dtype: int64

Producto con mayor ingreso generado:
nombre_producto
Vacuum-Pro-Black      1641168.16
Blender-XL-Red        1618005.99
Jacket-Winter-M       1612305.70
Sneakers-Urban-42     1570066.16
Laptop-Gaming-16GB    1089426.56
Name: monto_total, dtype: float64


In [65]:
# 4. Gasto en marketing por canal

In [66]:
gasto_por_canal = marketing1.groupby('canal')['gasto'].sum().sort_values(ascending=False)
print(gasto_por_canal)

# Con porcentaje del total
gasto_por_canal_pct = (gasto_por_canal / gasto_por_canal.sum() * 100).round(2)
resumen_canal = pd.DataFrame({
    'gasto_total': gasto_por_canal,
    'porcentaje': gasto_por_canal_pct
})
print(resumen_canal)

canal
social         976818.37
organic        972650.96
paid_search    922374.20
Name: gasto, dtype: float64
             gasto_total  porcentaje
canal                               
social         976818.37       34.01
organic        972650.96       33.87
paid_search    922374.20       32.12


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

La lectura comercial queda bien complementada con las métricas promedio y las agrupaciones.

Calculaste un ticket promedio de aproximadamente <b>$385.81</b>, una cantidad promedio de <b>1.50 productos</b> y analizaste los productos desde tres perspectivas: unidades, frecuencia de aparición e ingreso generado. También mostraste el gasto de marketing por canal y su participación porcentual.

Es especialmente positivo que diferencies “producto más vendido” de “producto con mayor ingreso”, porque ambas métricas responden preguntas distintas y pueden conducir a decisiones diferentes sobre inventario, promociones y rentabilidad.

</div>

---

## 🔹 Paso 3: Entender dónde se pierden los usuarios (funnel de conversión)

**🎯 Objetivo:** Analizar el comportamiento de los usuarios para identificar en qué etapa del proceso se pierden.


⚙️**Conexión a la base de datos**:  
Se ejecuta la línea de configuración para conectar con la base de datos y aplicar consultas SQL en la tabla **events**.

---

**📊 Parte 1: Construcción del funnel**
- ¿Cuántos usuarios llegan a cada etapa del funnel?  
- Se calcula el número de usuarios únicos por `nombre_evento`  
- Se ordenan los eventos según el flujo del usuario  

---

**📉 Parte 2: Análisis de conversión**
- Se calcula la tasa de conversión entre cada paso del funnel  
- Se identifica en qué etapa se pierde la mayor cantidad de usuarios  
- ¿Cuál es la tasa de conversión final?
---

In [108]:
import pandas as pd
from sqlalchemy import create_engine

# ======================
# Conexión (NO modificar)
# ======================
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-production-db-en'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [109]:
# Explorar tabla events
# =========================
query_events = '''
SELECT *
FROM events;
'''
events = pd.read_sql(query_events, con=engine)
events.head()

,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica


In [110]:
events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120000 entries, 0 to 119999
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   id_usuario          120000 non-null  object
 1   id_sesion           120000 non-null  object
 2   nombre_evento       120000 non-null  object
 3   timestamp_evento    120000 non-null  object
 4   pais                120000 non-null  object
 5   dispositivo         120000 non-null  object
 6   fuente_referencia   120000 non-null  object
 7   categoria_producto  120000 non-null  object
dtypes: object(8)
memory usage: 7.3+ MB


In [111]:
print(events.columns.tolist())
print(events.dtypes)

['id_usuario', 'id_sesion', 'nombre_evento', 'timestamp_evento', 'pais', 'dispositivo', 'fuente_referencia', 'categoria_producto']
id_usuario            object
id_sesion             object
nombre_evento         object
timestamp_evento      object
pais                  object
dispositivo           object
fuente_referencia     object
categoria_producto    object
dtype: object


In [112]:
print(events.columns.tolist())
events.head()

['id_usuario', 'id_sesion', 'nombre_evento', 'timestamp_evento', 'pais', 'dispositivo', 'fuente_referencia', 'categoria_producto']


,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica


In [113]:
# ======================
# PARTE 1: Totales del funnel
# ======================
query_totals = '''
SELECT 
    nombre_evento,
    COUNT(DISTINCT id_usuario) AS usuarios_unicos
FROM events
GROUP BY nombre_evento
ORDER BY 
    CASE nombre_evento
        WHEN 'first_visit' THEN 1
        WHEN 'select_item' THEN 2
        WHEN 'add_to_cart' THEN 3
        WHEN 'begin_checkout' THEN 4
        WHEN 'add_payment_info' THEN 5
        WHEN 'purchase' THEN 6
        ELSE 99
    END
'''
totals = pd.read_sql(query_totals, con=engine)
totals

,nombre_evento,usuarios_unicos
0,first_visit,7796
1,select_item,7582
2,add_to_cart,7634
3,begin_checkout,7208
4,add_payment_info,6250
5,purchase,6240


In [114]:
# ======================
# PARTE 2: Análisis de conversión
# ======================
# Tasa de conversión respecto a la primera etapa
totals['pct_del_total'] = (totals['usuarios_unicos'] / totals['usuarios_unicos'].iloc[0] * 100).round(2)

# Tasa de conversión etapa a etapa (drop-off)
totals['pct_etapa_anterior'] = (totals['usuarios_unicos'] / totals['usuarios_unicos'].shift(1) * 100).round(2)

# Caída de usuarios en número absoluto entre etapas consecutivas
totals['usuarios_perdidos'] = totals['usuarios_unicos'].shift(1) - totals['usuarios_unicos']

print(totals)

# Etapa donde se pierde más usuarios (mayor caída absoluta)
etapa_critica = totals.loc[totals['usuarios_perdidos'].idxmax(), 'nombre_evento']
print(f"\nEtapa con mayor pérdida de usuarios: {etapa_critica}")

# Tasa de conversión final (primera etapa → última etapa)
tasa_final = totals['usuarios_unicos'].iloc[-1] / totals['usuarios_unicos'].iloc[0] * 100
print(f"Tasa de conversión final (first_visit → purchase): {tasa_final:.2f}%")


      nombre_evento  usuarios_unicos  pct_del_total  pct_etapa_anterior  \
0       first_visit             7796         100.00                 NaN   
1       select_item             7582          97.26               97.26   
2       add_to_cart             7634          97.92              100.69   
3    begin_checkout             7208          92.46               94.42   
4  add_payment_info             6250          80.17               86.71   
5          purchase             6240          80.04               99.84   

   usuarios_perdidos  
0                NaN  
1              214.0  
2              -52.0  
3              426.0  
4              958.0  
5               10.0  

Etapa con mayor pérdida de usuarios: add_payment_info
Tasa de conversión final (first_visit → purchase): 80.04%


<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

El funnel cuenta usuarios únicos por nombre de evento y presenta las etapas en el orden esperado, pero todavía hay un problema importante en la forma de construir la conversión.

La consulta agrupa cada evento de manera independiente. Por esta razón aparecen <b>7,582</b> usuarios en <code>select_item</code> y <b>7,634</b> en <code>add_to_cart</code>, generando una conversión de <b>100.69%</b>. En un funnel secuencial, una etapa posterior no debería contener más usuarios que la etapa anterior, porque solo deberían avanzar quienes completaron previamente los pasos anteriores.

Adicionalmente, sin embargo, se esperaria una consulta estructurada mediante CTEs. y para mejorarlo debes reconstruir el funnel haciendo que cada etapa dependa de la anterior y verificando el orden temporal de los eventos del usuario. Después de ese ajuste, recalcula las tasas entre etapas e identifica nuevamente la caída principal.


</div>

---

## 🔹 Paso 4: Evaluar si los usuarios regresan (retención por cohortes)

**🎯 Objetivo:** Analizar la retención de usuarios para entender si regresan después de registrarse.

**Tablas**

- `users` 
- `user_activity` 

---
1. Se identifica la cohorte de cada usuario según el **mes de registro**.


2. Se calcula la retención semanal: cuántos usuarios **se mantienen activos** en cada semana desde su registro.
   - `retenido_w1`: usuarios activos en la semana 1  
   - `retenido_w2`: usuarios activos en la semana 2  
   - `retenido_w3`: usuarios activos en la semana 3  


3. Se calcula el porcentaje de retención para cada semana, dividiendo los usuarios retenidos entre los clientes iniciales de la cohorte:  
   - `semana_1`: porcentaje de usuarios retenidos en la semana 1  
   - `semana_2`: porcentaje de usuarios retenidos en la semana 2  
   - `semana_3`: porcentaje de usuarios retenidos en la semana 3  

Se revisa que la columna de fecha esté en formato correcto (`DATE`).  
Se realiza la conversión usando: `CAST(fecha_registro AS DATE)`

In [115]:
# Explorar tabla users
query_users = '''
SELECT 
    id_usuario,
    CAST(fecha_registro AS DATE) AS fecha_registro,
    país,
    dispositivo,
    tipo_plan
FROM users
LIMIT 5;
'''
users = pd.read_sql(query_users, con=engine)
print(users.dtypes)
users.head()

id_usuario        object
fecha_registro    object
país              object
dispositivo       object
tipo_plan         object
dtype: object


,id_usuario,fecha_registro,país,dispositivo,tipo_plan
0,user_0,2025-01-29,Mexico,mobile,free
1,user_1,2025-01-07,Mexico,mobile,free
2,user_2,2025-03-12,Argentina,mobile,free
3,user_3,2025-03-04,Mexico,desktop,free
4,user_4,2025-02-27,Argentina,desktop,free


In [116]:
users['fecha_registro'] = pd.to_datetime(users['fecha_registro'])
print(users.dtypes)

id_usuario                object
fecha_registro    datetime64[ns]
país                      object
dispositivo               object
tipo_plan                 object
dtype: object


In [117]:
# Explorar tabla user_activity
query_activity = '''
SELECT *
FROM user_activity
LIMIT 5;
'''
user_activity = pd.read_sql(query_activity, con=engine)
user_activity['fecha_actividad']=pd.to_datetime(user_activity['fecha_actividad'])
print(user_activity.dtypes)
user_activity.head()

id_usuario                       object
fecha_actividad          datetime64[ns]
dias_despues_registro             int64
activo                            int64
dtype: object


,id_usuario,fecha_actividad,dias_despues_registro,activo
0,user_0,2025-02-05,7,0
1,user_0,2025-02-12,14,1
2,user_0,2025-02-19,21,1
3,user_0,2025-02-26,28,0
4,user_1,2025-01-14,7,0


In [148]:
# Retención por cohortes
# ======================
query_retencion = '''
WITH usuarios_cohorte AS (
    SELECT
        id_usuario,
        DATE_TRUNC('month', CAST(fecha_registro AS DATE)) AS cohorte
    FROM users
),
actividad_semanal AS (
    SELECT
        id_usuario,
        MAX(CASE WHEN dias_despues_registro = 7  THEN activo ELSE 0 END) AS retenido_w1,
        MAX(CASE WHEN dias_despues_registro = 14 THEN activo ELSE 0 END) AS retenido_w2,
        MAX(CASE WHEN dias_despues_registro = 21 THEN activo ELSE 0 END) AS retenido_w3
    FROM user_activity
    GROUP BY id_usuario
)
SELECT
    uc.cohorte,
    COUNT(DISTINCT uc.id_usuario) AS clientes_iniciales,
    SUM(COALESCE(a.retenido_w1, 0)) AS retenido_w1,
    SUM(COALESCE(a.retenido_w2, 0)) AS retenido_w2,
    SUM(COALESCE(a.retenido_w3, 0)) AS retenido_w3
FROM usuarios_cohorte uc
LEFT JOIN actividad_semanal a ON uc.id_usuario = a.id_usuario
GROUP BY uc.cohorte
ORDER BY uc.cohorte
'''

retencion = pd.read_sql(query_retencion, con=engine)
retencion

,cohorte,clientes_iniciales,retenido_w1,retenido_w2,retenido_w3
0,2025-01-01 00:00:00+00:00,1627,697.0,668.0,656.0
1,2025-02-01 00:00:00+00:00,1444,611.0,609.0,635.0
2,2025-03-01 00:00:00+00:00,1636,677.0,705.0,690.0
3,2025-04-01 00:00:00+00:00,1606,680.0,697.0,663.0
4,2025-05-01 00:00:00+00:00,1687,695.0,676.0,706.0


In [119]:

#Calculo de porcentaje de retencion
retencion['semana_1'] = (retencion['retenido_w1'] / retencion['clientes_iniciales'] * 100).round(2)
retencion['semana_2'] = (retencion['retenido_w2'] / retencion['clientes_iniciales'] * 100).round(2)
retencion['semana_3'] = (retencion['retenido_w3'] / retencion['clientes_iniciales'] * 100).round(2)

retencion


,cohorte,clientes_iniciales,retenido_w1,retenido_w2,retenido_w3,semana_1,semana_2,semana_3
0,2025-01-01 00:00:00+00:00,1627,697.0,668.0,656.0,42.84,41.06,40.32
1,2025-02-01 00:00:00+00:00,1444,611.0,609.0,635.0,42.31,42.17,43.98
2,2025-03-01 00:00:00+00:00,1636,677.0,705.0,690.0,41.38,43.09,42.18
3,2025-04-01 00:00:00+00:00,1606,680.0,697.0,663.0,42.34,43.40,41.28
4,2025-05-01 00:00:00+00:00,1687,695.0,676.0,706.0,41.20,40.07,41.85


<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

La cohorte mensual está correctamente construida desde <code>fecha_registro</code> y los porcentajes utilizan como denominador los clientes iniciales de cada cohorte. Sin embargo, la lógica de retención todavía no coincide con la solicitada en la rúbrica.

Actualmente utilizas condiciones exactas para los días 7, 14 y 21:

<code>dias_despues_registro = 7</code>,  
<code>dias_despues_registro = 14</code> y  
<code>dias_despues_registro = 21</code>.

Para mejorarlo, se esperaria convertir los días posteriores al registro en semanas mediante una lógica equivalente a <code>FLOOR(dias_despues_registro / 7)</code> y calcular retención acumulada usando condiciones como semana mayor o igual a 1, 2 y 3.

La diferencia es importante: tu consulta actual mide actividad en un día o semana exacta, mientras que la retención acumulada busca saber cuántos usuarios continuaron activos desde determinada semana en adelante. Ajusta primero la variable de semana y después recalcula <code>retenido_w1</code>, <code>retenido_w2</code>, <code>retenido_w3</code> y sus porcentajes.

</div>

## 🔹 Paso 5: Validar si los cambios generan impacto (test estadístico)

🎯 **Objetivo:** Evaluar si la modificación en la UI del checkout impacta la **tasa de conversión de compra**.

---

1. **Analizar el dataset** `experiment_checkout_ui.csv` para identificar la métrica principal **conversion**.
   - La métrica **conversion** es 1 si el usuario completó la compra, 0 si no.    
2. **Plantear la hipótesis estadística**     
3. **Aplicar el test estadístico adecuado** 
4. **Interpretar el resultado**  

---
Hipótesis estadística
   - **H₀ (Hipótesis nula):** No existe diferencia entre la tasa de conversion del grupo de control y de tratamiento
   - **H₁ (Hipótesis alternativa):** Existe una diferencia en la tasa de conversion entre ambos grupos
   
**Test estadístico:** .z  
**Nivel de significancia alpha:** 0.05

In [120]:
experiment_checkout_ui=pd.read_csv('experiment_checkout_ui.csv')

In [121]:
experiment_checkout_ui.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id_usuario       10000 non-null  object 
 1   variante         10000 non-null  object 
 2   convirtio        10000 non-null  int64  
 3   dispositivo      10000 non-null  object 
 4   pais             10000 non-null  object 
 5   duracion_sesion  10000 non-null  float64
 6   timestamp        10000 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 547.0+ KB


In [122]:
# Paso 1: Explorar el dataset y confirmar la métrica
print(experiment_checkout_ui.shape)
print(experiment_checkout_ui.dtypes)
experiment_checkout_ui.head()

# Confirmar los grupos experimentales
print(experiment_checkout_ui['variante'].unique())


(10000, 7)
id_usuario          object
variante            object
convirtio            int64
dispositivo         object
pais                object
duracion_sesion    float64
timestamp           object
dtype: object
['tratamiento' 'control']


In [123]:
# Aplicacion del test estadistico
from statsmodels.stats.proportion import proportions_ztest

# Tasa de conversión por grupo
resumen = experiment_checkout_ui.groupby('variante')['convirtio'].agg(
    conversiones='sum', usuarios='count', tasa_conversion='mean'
)
print(resumen)

# Preparar datos para el test
control = experiment_checkout_ui[experiment_checkout_ui['variante'] == 'control']['convirtio']
tratamiento = experiment_checkout_ui[experiment_checkout_ui['variante'] == 'tratamiento']['convirtio']

count = [tratamiento.sum(), control.sum()]   # conversiones exitosas por grupo
nobs = [len(tratamiento), len(control)]      # tamaño de cada grupo

z_stat, p_value = proportions_ztest(count, nobs, alternative='two-sided')

print(f"\nZ-statistic: {z_stat:.4f}")
print(f"p-value: {p_value:.4f}")

             conversiones  usuarios  tasa_conversion
variante                                            
control               779      4965         0.156898
tratamiento           820      5035         0.162860

Z-statistic: 0.8133
p-value: 0.4161


In [124]:
#  Interpretar el resultado
alpha = 0.05
if p_value < alpha:
    print(f"Se rechaza H0 (p={p_value:.4f} < {alpha}): la diferencia en conversión es estadísticamente significativa.")
else:
    print(f"No se rechaza H0 (p={p_value:.4f} >= {alpha}): no hay evidencia suficiente de diferencia en conversión.")

No se rechaza H0 (p=0.4161 >= 0.05): no hay evidencia suficiente de diferencia en conversión.


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Excelente trabajo en el análisis del experimento A/B.

Definiste correctamente una hipótesis nula de ausencia de diferencia y una hipótesis alternativa de existencia de diferencia entre las tasas de conversión. Luego agrupaste conversiones y usuarios por variante, preparaste correctamente <code>count</code> y <code>nobs</code>, aplicaste <code>proportions_ztest</code> y comparaste el p-valor contra <code>alpha = 0.05</code>.

La conclusión también utiliza un lenguaje estadístico apropiado: con un p-valor cercano a <b>0.4161</b>, no se rechaza la hipótesis nula y se indica que no existe evidencia suficiente para afirmar una diferencia en conversión. Evitaste decir que las tasas son exactamente iguales, lo cual demuestra una buena interpretación.

</div>

---

## 🔹 Paso 6: Comunicar los resultados (Dashboard en BI)

🎯 **Objetivo**:  
Crear un dashboard que muestre de manera clara y visual los resultados del análisis de ventas, costos, marketing y conversión. 

Se usarán los CSVs limpios del Paso 1:

- `orders_clean.csv`  
- `catalog_clean.csv`  
- `marketing_clean.csv`

---

1️⃣ Preparación de los datos
1. Cargar los CSVs en Power BI o Tableau.
2. Revisar relaciones:
   - `orders.nombre_producto` → `catalog.nombre_producto`
   - `orders.fecha_pedido` → tabla de fechas (crear calendario para análisis temporal)
   - `orders.fecha_pedido` → `dim_fecha.date`
3. Crear columnas calculadas necesarias
4. Crear tabla de fechas para poder calcular comparaciones YTD, YoY o períodos anteriores (`Previous Year`, `Previous Month`).

---

2️⃣ Dashboard 1: Overview Ejecutivo
**KPIs principales a mostrar:**
- Revenue total
- Profit total
- Gasto total en marketing
- Ticket promedio
- Cantidad promedio de productos por orden

**Visualizaciones sugeridas:**
- Tarjetas KPI para revenue, profit y gasto marketing
- Gráfico de líneas: evolución mensual de revenue o profit
- Gráfico de líneas YTD
- Gráfico de barras: revenue y profit por producto o categoría

---

 3️⃣ Dashboard 2: Detalle / Drill-through  
**Objetivo:** Permitir explorar los datos desde el KPI general hasta cada orden o producto.

**Visualizaciones sugeridas:**
- Tabla detallada de órdenes con:
  - producto, cantidad, revenue, cost, profit
  - color condicional (profit negativo en rojo, positivo en verde)
- Gráfico de barras por producto con medida `cantidad vendida`
- Drill-through: seleccionar un producto y ver todos los pedidos relacionados
- Filtros por fecha, categoría de producto, etc

---

## 🚀 Entrega Final

Comparte el acceso a tu Dashboard para revisión.   
Puedes entregar el Dashboard utilizando **Power BI o Tableau**.

Incluye **uno de los siguientes**:

- 🔗 Link público del dashboard publicado en **Power BI Service o Tableau Public / Tableau Cloud**
- 🔗 Link de **Google Drive o OneDrive** con el archivo del proyecto (`.pbix`) y los 3 csvs limpios.


### 📎 Enlace del Dashboard

In [67]:
# https://drive.google.com/drive/folders/1QMZjWcNvi9RyH6VQ0ITgizrbmALKnLQI?usp=sharing


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a><br><br>

Se reconoce un buen esfuerzo en la construcción del dashboard.

La página de resumen incluye tarjetas para Revenue, Profit, Marketing, Ticket promedio, ventas MTD y ventas YTD. También presenta evolución mensual, comparación por categoría, una tabla de detalle y filtros por producto. La segunda página permite profundizar mediante categoría, año-mes y producto, además de incorporar formato condicional en la tabla.

La estructura de resumen a detalle cumple con una lógica útil para un usuario de negocio y demuestra que trabajaste no solo el análisis técnico, sino también su comunicación visual.

</div>